pkg/webhook/policy.go

```diff
@@ -342,18 +344,33 @@ func (h *validationHandler) validateGatekeeperResources(req *admission.Request)
 // Returns an error if the ConstraintTemplate fails validation.
 // The returned boolean is only true if error is non-nil and is a result of user
 // error.
-func (h *validationHandler) validateTemplate(req *admission.Request) (bool, error) {
+func (h *validationHandler) validateTemplate(ctx context.Context, req *admission.Request) (bool, error) {
 	templ, _, err := deserializer.Decode(req.AdmissionRequest.Object.Raw, nil, nil)
 	if err != nil {
 		return false, err
 	}
 
 	unversioned := &templates.ConstraintTemplate{}
-	if err := runtimeScheme.Convert(templ, unversioned, nil); err != nil {
+	err = runtimeScheme.Convert(templ, unversioned, nil)
+	if err != nil {
 		return false, err
 	}
 
-	if err := h.opa.ValidateConstraintTemplate(unversioned); err != nil {
+	// Ensure that it is possible to generate a CRD for this ConstraintTemplate.
+	_, err = h.opa.CreateCRD(unversioned)
+	if err != nil {
+		return true, err
+	}
+
+	// Create a temporary Driver and attempt to add the Template to it. This
+	// ensures the Rego code both parses and compiles.
+	d, err := local.New()
+	if err != nil {
+		return false, fmt.Errorf("unable to create Driver: %v", err)
+	}
+
+	err = d.AddTemplate(ctx, unversioned)
```

Comment:
Do we want to block a response on compiling a template? That could be an expensive/long-running operation.

Reply:
There could be a scaling bottleneck here if there were multiple simultaneous inbound requests, but we can see how that goes, I think.